README.md
CPU 6502 Simulator
==================

`cpu-6502-simulator` is a complete MOS 6502 CPU simulator entirely written in [Scala](http://www.scala-lang.org/). The goal of this project has been to provide a tool for execution of any 6502-compiled code in a simulation-based environment. It was heavily inspired and is primarily based upon the work done by _Ullrich von Bassewitz_ in his `sim65` program (originally written in plain `C`).

### What was the motivation behind building another MOS 6502 CPU simulator?

1. There has been no such utility written in [Scala](http://www.scala-lang.org/) ever before.
2. Just to treat it as a practice in building [Scala](http://www.scala-lang.org/) applications.
3. As a necessity is the mother of all invention, and since back then I was aiming at developing a complete [unit-testing framework](https://github.com/pawelkrol/Scala-CommTest) targetting MOS 6502 CPU (specifically any code that was meant to run on the [Commodore 64](https://en.wikipedia.org/wiki/Commodore_64)) to help me with testing of complex subroutine executions, I considered writing this simulator an interim step of a larger development process (while still designed to be usable as a standalone tool, this simulator was not an ultimate goal of my work). Another potential utility of this library is to ease implementation of a 6502 disassembler of binary programs.

`cpu-6502-simulator` is not restricted to be executed as a standalone application JAR, although it leaves this possibility (without too many options to configure simulation environment however). It comes up bundled with the minimal set of dependencies required to build, test and include it in your own [Scala](http://www.scala-lang.org/) projects.

PREREQUISITES
-------------

This `cpu-6502-simulator` application is setup with [sbt 0.13.17](http://www.scala-sbt.org/) as a build tool, [sbt-proguard 0.3.0](https://github.com/sbt/sbt-proguard) as an sbt [plugin](http://www.scala-sbt.org/0.13/docs/Plugins.html) for running [ProGuard 5.3.3](http://proguard.sourceforge.net/), [ScalaTest 3.0.5](http://www.scalatest.org/) as a unit-testing framework, [Scalactic 3.0.5](http://www.scalactic.org/) as a small library of utilities, [scopt 3.7.0](https://github.com/scopt/scopt) as a command line options parsing library, [Scala Logging 3.8.0](https://github.com/typesafehub/scala-logging) as a logging library wrapping [SLF4J](http://www.slf4j.org/), and [Logback 1.2.3](http://logback.qos.ch/) as a backend logging framework.

Dependency management is normally handled automatically by your build tool.

`cpu-6502-simulator` may optionally be packaged as a standalone executable JAR using `proguard:proguard` command of an `sbt` shell.

VERSION
-------

Version 0.05-SNAPSHOT (2018-02-27)

INSTALLATION
------------

You can automatically download and install this library by adding the following dependency information to your `build.sbt` configuration file:

    resolvers += "Sonatype Snapshots" at "https://oss.sonatype.org/content/repositories/snapshots/"

    libraryDependencies += "com.github.pawelkrol" % "cpu-6502-simulator" % "0.05-SNAPSHOT"

In order to build and run an executable application JAR type the following:

    $ git clone git://github.com/pawelkrol/cpu-6502-simulator.git
    $ cd cpu-6502-simulator/
    $ sbt clean update compile test package proguard:proguard

    $ java -Dfile.encoding=UTF8 -jar target/proguard/cpu-6502-simulator-0.05-SNAPSHOT.jar

You may have to add the following automatic export to your `~/.bash_profile`:

    export _JAVA_OPTIONS="-Xms1024m -Xmx2G -Xss256m"

In order to publish an artifact to [Sonatype](https://oss.sonatype.org/) type the following:

    $ sbt publishSigned publishLocal

EXAMPLES
--------

Load program from a file on disk:

    import com.github.pawelkrol.CPU6502.ByteVal
    import com.github.pawelkrol.CPU6502.Core

    import io.Codec.ISO8859
    import io.Source.fromFile

    val core = Core()

    val fileName = "main.prg"
    val targetAddress = 0x1000.toShort

    val data = fromFile(fileName)(ISO8859).toArray.map(ByteVal(_))
    core.memory.write(targetAddress, data)

Write compiled code directly into a memory:

    import com.github.pawelkrol.CPU6502.Core

    val core = Core()

    val targetAddress = 0x1000.toShort

    /** LDY #$00
     *  LDA ($FB),Y
     *  STA ($FD),Y
     *  RTS
     */
    core.memory.write(targetAddress, 0xa0, 0x00, 0xb1, 0xfb, 0x91, 0xfd, 0x60)

Run simulator until 100 cycles count is reached:

    import com.github.pawelkrol.CPU6502.Core
    import com.github.pawelkrol.CPU6502.Runner

    import java.io.File

    val core = Core()
    core.reset
    core.register.PC = 0x1000

    val cycleCount = 100
    val file = new File("main.prg")

    Runner.go(core, file, Some(cycleCount))

Apply the same command that additionally produces verbose output:

    import com.github.pawelkrol.CPU6502.Application
    import com.github.pawelkrol.CPU6502.Core
    import com.github.pawelkrol.CPU6502.Runner

    import java.io.File

    val core = Core()
    core.reset
    core.register.PC = 0x1000

    val cycleCount = 100
    val file = new File("main.prg")

    Application.verbose = true

    Runner.go(core, file, Some(cycleCount))

Simulate running a program using +60k RAM memory expansion:

    import com.github.pawelkrol.CPU6502.Commodore.ExtendedMemory
    import com.github.pawelkrol.CPU6502.Core

    val memory = ExtendedMemory()
    val core = Core(memory)

    /** Use the following switches in a source code of a simulated program:
     *
     *  ; Switch $1000-$ffff address space to an additional memory bank:
     *  LDY #$80
     *  STA $D100
     *
     *  ; Switch $1000-$ffff address space to an onboard memory bank:
     *  LDY #$00
     *  STA $D100
     */

    // Runner.go(core, file, Some(cycleCount))

COPYRIGHT AND LICENCE
---------------------

Copyright (C) 2016-2018 by Pawel Krol.

This library is free open source software; you can redistribute it and/or modify it under [the same terms](https://github.com/pawelkrol/cpu-6502-simulator/blob/master/LICENSE.md) as Scala itself, either Scala version 2.12.4 or, at your option, any later version of Scala you may have available.


CHANGES.md
CHANGES
=======

Revision history for `cpu-6502-simulator`, a complete MOS 6502 CPU simulator entirely written in [Scala](http://www.scala-lang.org/).

0.05-SNAPSHOT (2018-02-27)
--------------------------

* TODO

0.04 (2018-02-27)
-----------------

* Enhancement: Expose direct access to Charset Generator and VIC data
* Enhancement: Ensure VIC data of new `CommodoreMemory` instances is always initialised with Kernal's I/O defaults
* `scala-logging` version upgraded to 3.8.0

0.03 (2018-02-07)
-----------------

* Enhancement: Enable customisation of a simulated program's memory by replacing `Memory` class with a generic trait that a user's program may extend from
* New feature: Provide an implementation of `CommodoreMemory` which may be used as a replacement of `Memory` object specified upon `Core` initialisation (it uses Kernal, Charset Generator, and Basic binaries, selecting an appropriate memory module based on a value change applied to `$01` I/O register)
* New feature: Provide an implementation of `ExtendedMemory` which is capable of simulating +60k RAM memory expansion
* `Scala` version upgraded to 2.12.4

0.02 (2017-01-01)
-----------------

* Refactoring: Allow instantiation of multiple `Core` objects by removing all references to a single `Core` instance accessible globally in the main `Application` object (previous implementation resulted in inability to instantiate a standalone `Core` object and its tight coupling with the main `Application` object)
* Bug fixed: Add 1 to program counter pulled from stack when returning from subroutine (previous implementation resulted in broken RTS operation)
* Bug fixed: Repair broken packaging while publishing artifact to remote repository (previous implementation resulted in creating application and javadoc JARs overwritten with contents of sources JAR)
* `Scala` version upgraded to 2.12.1

0.01 (2016-11-16)
-----------------

* Initial version (supports all documented MOS 6502 opcodes, provides step-by-step execution of an assembled source code, exposes core system with a direct access to all CPU registers and an entire addressable memory, and implements experimental command line simulator enabling execution of an arbitrary program file with a 6502-compiled code)


src/main/scala/com/github/pawelkrol/CPU6502/Application.scala
package com.github.pawelkrol.CPU6502

import java.io.File
import scopt.OptionParser

object Application extends Logging {

  private val appVersion = "0.05-SNAPSHOT"

  private val core = Core()

  private val parser = new OptionParser[Arguments]("cpu-6502-simulator") {
    head("cpu-6502-simulator", appVersion)
    help("help") text("prints out this usage text")
    opt[Unit]("verbose") optional() action { (_, option: Arguments) =>
      option.copy(verbose = true) } text("produces verbose output, including all CPU instructions that are executed")
    opt[Int]("cycle-count") optional() action { (value: Int, option: Arguments) =>
      option.copy(cycleCount = Some(value)) } text("instructs program to exit simulator after a given number of cycles")
    opt[String]("start-address") optional() action { (value: String, option: Arguments) =>
      option.copy(startAddress = Some(Integer.parseInt(value, 16))) } text("overwrites the default (i.e., $0200) start address, must be given as a number consisting of 4 hexadecimal digits (e.g., \"c000\")")
    arg[File]("<filename>") unbounded() required() action { (file: File, option: Arguments) =>
      option.copy(file = Some(file)) } text("program file to execute (e.g., \"main.prg\")")
  }

  def main(args: Array[String]) = {
    println("\nCPU 6502 Simulator %s (2018-02-27)\nCopyright (C) 2016-2018 Pawel Krol (DJ Gruby/Protovision/TRIAD)\n".format(appVersion))

    parser.parse(args, Arguments()) match {
      case Some(arguments) => {
        try {
          arguments.validate
        } catch {
          case e: Exception => {
            println(e.getMessage)
            System.exit(1)
          }
        }
        runWith(arguments)
      }
      case None =>
        // arguments are invalid, usage message will have been displayed
    }

    System.exit(0)
  }

  def runWith(arguments: Arguments) {
    core.reset

    // Initialize program counter with a custom start address
    core.register.PC = arguments.startAddress match {
      case Some(addr) =>
        addr.toShort
      case None =>
        core.memory.get_val_from_addr(0xfffc.toShort)
    }

    verbose = arguments.verbose

    Runner.go(core, arguments.file.get, arguments.cycleCount)
  }
}
